### DISCUSSION TITLE
Can't import fiona because `libgdal.29.dylib` can't be found

### DISCUSSION THREAD

**Discussion by ischmidt20** [Toblerity/Fiona#1131] [Q&A] [answered]
## Expected behavior and actual behavior.

`import fiona` should import the `fiona` module.

Instead, I see the following:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/lib/python3.9/site-packages/fiona/__init__.py", line 86, in <module>
    from fiona.collection import BytesCollection, Collection
  File "/opt/homebrew/lib/python3.9/site-packages/fiona/collection.py", line 11, in <module>
    from fiona.ogrext import Iterator, ItemsIterator, KeysIterator
ImportError: dlopen(/opt/homebrew/lib/python3.9/site-packages/fiona/ogrext.cpython-39-darwin.so, 0x0002): Library not loaded: /opt/homebrew/opt/gdal/lib/libgdal.29.dylib
  Referenced from: /opt/homebrew/lib/python3.9/site-packages/fiona/ogrext.cpython-39-darwin.so
  Reason: tried: '/opt/homebrew/opt/gdal/lib/libgdal.29.dylib' (no such file), '/usr/local/lib/libgdal.29.dylib' (no such file), '/usr/lib/libgdal.29.dylib' (no such file), '/opt/homebrew/Cellar/gdal/3.4.1_1/lib/libgdal.29.dylib' (no such file), '/usr/local/lib/libgdal.29.dylib' (no such file), '/usr/lib/libgdal.29.dylib' (no such file)
```

## Operating system

M1 Mac running macOS Monterey 12.1

## Fiona and GDAL version and provenance

Fiona version 1.8.20 installed via pip

GDAL version 3.4.1_1 installed with brew

I "fixed" the issue simply by running the following command:

`cp /opt/homebrew/Cellar/gdal/3.4.1_1/lib/libgdal.30.dylib /opt/homebrew/Cellar/gdal/3.4.1_1/lib/libgdal.29.dylib`

But I thought I would report a bug because I couldn't find any information on this issue, and the "solution" is not obvious.

**Comment by sgillies** [comment]
@ischmidt20 I'm sorry you're having a hard time. The line `ImportError: dlopen(/opt/homebrew/lib/python3.9/site-packages/fiona/ogrext.cpython-39-darwin.so, 0x0002): Library not loaded: /opt/homebrew/opt/gdal/lib/libgdal.29.dylib` suggests to me that your fiona installation was built based on an older version of GDAL that no longer exists on your computer. Instead of copying the shared library, I recommend reinstalling the fiona package.

**Comment by krober10nd** [comment]
I faced the same issue. I had to rebuild from source. Reinstalling via pip did not avoid the error.

**Comment by nicolegoebel** [comment]
renaming libgdal.30.dylib to libgdal.29.dylab worked for me (but my files were in `/usr/local/opt/gdal/lib/`

**Comment by sgillies** [comment]
@nicolegoebel renaming the libraries will almost certainly break something. I can't recommend anything other than what @krober10nd has done. Install GDAL using homebrew, make sure that gdal-config is on your executable path, and then `pip install --no-binary=fiona fiona`.

**Comment by nicolegoebel** [comment]
Okay thanks!!

On Tue, Jun 7, 2022 at 1:48 PM Sean Gillies ***@***.***>
wrote:

> @nicolegoebel <https://github.com/nicolegoebel> renaming the libraries
> will almost certainly break something. I can't recommend anything other
> than what @krober10nd <https://github.com/krober10nd> has done. Install
> GDAL using homebrew, make sure that gdal-config is on your executable path,
> and then pip install --no-binary=fiona fiona.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Toblerity/Fiona/issues/1054#issuecomment-1149152969>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANPPCSZYALVSVMHW7EFBODVN6YTTANCNFSM5MLWPVUA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Nicole Goebel

Data Scientist


ACLIMA INC

Direct (415) 735-5389 |  Email ***@***.***

10 Lombard St. Suite 400  |  San Francisco, CA 94111

Hello, we're Aclima <http://www.aclima.io/>.

This email and any attachments may contain private, confidential and
privileged material for the sole use of the intended recipient. If you are
not the intended recipient, please immediately delete this email and any
attachments.

**Comment by nicolegoebel** [comment]
> 

Actually this did not work for me and now I am back to where I started. Except that there is no longer a `libgdal.XX.dylib` file. Can you make any recommendations to fix this?? THanks!

**Comment by ischmidt20** [comment]
I also want to chime in and say that reinstalling gdal and fiona with various combinations of arguments never fixed this. The only thing that worked was simply renaming the file as described above, which I'll admit was a very surprising solution.

**Comment by rickemery** [comment]
I am having having this issue as well, but I have `libgdal.31.dylib` and am missing `libgdal.30.dylib`. No reinstalling has worked for me. I had to rename the files as mentioned above.

**Comment by sgillies** [comment, marked as answer]
Update: I've got a hypothesis at https://github.com/Toblerity/Fiona/discussions/1131#discussioncomment-4822803. Please read it before you rename your library.

To be clear: the project does not endorse renaming the shared library file. It's likely to trip you up again in the future.

I've converted this to a discussion and hope that someone who knows homebrew inside and out can explain how to get homebrew libgdal into good working order.

**Comment by jameshalgren** [comment]
The same issue cropped up with fiona 1.18 and libgdal.32 on macOS with brew installed gdal at version 3.6.2. The same correction -- copying (I just created a symlink...) the file to create the path to the sought-for but missing earlier version. 
```
(pyenv310) % pip install --force fiona
Collecting fiona
  Using cached Fiona-1.8.22-cp310-cp310-macosx_12_0_arm64.whl
  ...
  Attempting uninstall: fiona
    Found existing installation: Fiona 1.8.22
    Uninstalling Fiona-1.8.22:
      Successfully uninstalled Fiona-1.8.22
(pyenv310) % python test_script_importing_fiona.py
Traceback (most recent call last):
  ...
ImportError: dlopen(/Users/user/pyenv310/lib/python3.10/site-packages/fiona/ogrext.cpython-310-darwin.so, 0x0002): Library not loaded: '/opt/homebrew/opt/gdal/lib/libgdal.31.dylib'
  Referenced from: '/Users/user/pyenv310/lib/python3.10/site-packages/fiona/ogrext.cpython-310-darwin.so'
  Reason: tried: '/opt/homebrew/opt/gdal/lib/libgdal.31.dylib' (no such file), '/usr/local/lib/libgdal.31.dylib' (no such file), '/usr/lib/libgdal.31.dylib' (no such file), '/opt/homebrew/Cellar/gdal/3.6.2_2/lib/libgdal.31.dylib' (no such file), '/usr/local/lib/libgdal.31.dylib' (no such file), '/usr/lib/libgdal.31.dylib' (no such file)
(pyenv310) % ls /opt/homebrew/Cellar/gdal/3.6.2_2/lib/libgdal.*
/opt/homebrew/Cellar/gdal/3.6.2_2/lib/libgdal.32.3.6.2.dylib
(pyenv310) % ln -s /opt/homebrew/Cellar/gdal/3.6.2_2/lib/libgdal.32.3.6.2.dylib /opt/homebrew/Cellar/gdal/3.6.2_2/lib/libgdal.31.dylib
(pyenv310) % python test_script_importing_fiona.py
running script ...

```

**Reply by seanrichardson95** [reply]
Ran into this too. I just renamed my file to `.31.` and things seem to be working (for now)

**Reply by sgillies** [reply]
@ischmidt20 @seanrichardson95 I think with the 1.9.0 release done I was able to free up some brain cells to think harder about this. I have a hypothesis.

There are no macosx arm wheels for 1.8.22 on PyPI. Thus `Using cached Fiona-1.8.22-cp310-cp310-macosx_12_0_arm64.whl` tells me that this wheel was built from source and saved the last time you ran `pip install fiona` on your computer. It links libgdal.31.dylib and I believe that is because your homebrew GDAL version then was 3.5.

My hypothesis continues: since, you've upgraded to 3.6.2. libgdal.31.dylib is gone (replaced by libgdal.32.dylib), but that wheel still needs it. And since pip can't tell that your cached wheel is incompatible with GDAL 3.6, it doesn't know that you need to pass over the cached wheel and make a completely new one.

Can you try something for me please? Remove your symlinks or restore the .dylib file to its true name. Then run and test the following.

```
pip install --force-reinstall --no-cache-dir fiona
```

This will ignore cached wheels. I'm eager to hear what happens.

Afterwards, you might want to remove all the old cached wheels.
 
```
pip cache remove Fiona*
```

**Reply by seanrichardson95** [reply]
Sorry @sgillies , I just finally got server at my work to actually run on my local environment so I'm not going to do anything to possibly mess that up. If this was for a personal project I'd give it a try tho!

**Reply by sgillies** [reply]
@seanrichardson95 I understand. Keep my comment in mind in case you ever upgrade homebrew on your server again.

**Reply by ischmidt20** [reply]
@sgillies So I uninstalled and reinstalled gdal to ensure a clean installation with the latest version. Then I reinstalled fiona with pip via your command above. I opened a new python session and was able to `import fiona` with no errors! I still have to play around to make sure everything works as expected, but I think your solution may have done the trick. Thanks!

**Comment by marianhlavac** [comment]
Same issue here, running commands as suggested in https://github.com/Toblerity/Fiona/discussions/1131#discussioncomment-4822803 did not help.

**Reply by jameshalgren** [reply]
This issue cropped up again and required the same fix. You can see comparing the brew path below with my previous comment that I've updated gdal somewhere along the way (there was a reason for that but I don't recall exactly). When I returned to work on the fiona-related branch, the error had returned. 

Unfortunately, this didn't seem to fix it: 
``` sh
pip install --force-reinstall --no-cache-dir fiona
```

Maybe I missed a step?

This fixed it, like before. 

``` sh
(pyenv310) % ln -s /opt/homebrew/Cellar/gdal/3.6.2_3/lib/libgdal.32.3.6.2.dylib /opt/homebrew/Cellar/gdal/3.6.2_3/lib/libgdal.31.dylib
```

**Reply by sgillies** [reply]
I'm beginning to suspect that nothing short of a fiona homebrew formula will work. If you're using homebrew GDAL and homebrew Python, might as well get Fiona the same route.

A reminder: renaming your shared library files is a terrible hack.
